### PR TITLE
#442: Improves documentation and the output error message when /noVerify is used

### LINKED ISSUE #442: Documentation issue: surprising that /noVerify blocks /compile and /spillTargetCode
Some combinations of these flags results in the Dafny CLI failing to output CS code:

```
> echo "method Main() {}" >> test.dfy

> dafny test.dfy
Dafny 2.3.0.10506

Dafny program verifier finished with 0 verified, 0 errors
Compiled assembly into test.exe

> dafny test.dfy /noVerify
Dafny 2.3.0.10506

Dafny program verifier finished with 0 verified, 0 errors

> dafny test.dfy /compile:0 /spillTargetCode:2
Dafny 2.3.0.10506

Dafny program verifier finished with 0 verified, 0 errors
Compiled program written to test.cs

dafny test.dfy /compile:0 /spillTargetCode:2 /noVerify
Dafny 2.3.0.10506

Dafny program verifier finished with 0 verified, 0 errors
```

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #442

### CODE REVIEW

**Comment by robin-aws** [comment]
I definitely like the options documentation improvement - thanks!

For the Boogie trailer message, the new version is an improvement, but still potentially confusing since the verifier didn't "finish" doing anything in the first place. I'd prefer to not see the message at all if /noVerify was specified (i.e. just don't call WriteTrailer) - AFAICT it will never print anything interesting if Boogie wasn't invoked. I prefer that to duplicating code.

While we're at it... :) can we add a `/verifyVerbose:0` flag to turn this off, similar to the `/compileVerbose:0` flag? That would reduce a lot of cruft in #680, since we can avoid having .expect files for a huge number of test cases that only assert verification passes successfully.

**Comment by davidcok** [comment]
Well, the message is ‘Dafny finished without attempting verification’, which is entirely accurate — Dafny did parsing, type resolution and related type checks, and translation to ppl, everything except verifying the ppl. I prefer having the message there (e.g. so you know that Dafny did not abort in the middle of its work). I could live with a /verifyVerbose flag if you and Rustan think it is needed.

As for test cases, I would definitely prefer that the message be present - again to distinguish complete, correct behavior from aborted behavior.

- David

**Comment by davidcok** [comment]
Instead of adding /verifyVerbose, could there be an optional argument to /noVerify ?

**Comment by robin-aws** [comment]
> Well, the message is ‘Dafny finished without attempting verification’, which is entirely accurate — Dafny did parsing, type resolution and related type checks, and translation to ppl, everything except verifying the ppl. I prefer having the message there (e.g. so you know that Dafny did not abort in the middle of its work). I could live with a /verifyVerbose flag if you and Rustan think it is needed. 

The message says "Dafny program verifier finished..." rather than just "Dafny finished...", so I stand by the statement that its still a bit confusing. :) We could tweak it to fix that though. I wouldn't use "Dafny finished" since it raise questions about what Dafny refers to at runtime - does compiling and/or running the program (which will using the Dafny runtime) also count as "Dafny"?

> As for test cases, I would definitely prefer that the message be present - again to distinguish complete, correct behavior from aborted behavior.

I'd rather not have that principle, since it's a slippery slope to way too much output - it implies we shouldn't be using /compileVerbose:0 in tests either, and perhaps should be printing out stuff like "Resolution finished successfully". I'm willing to defer to stronger opinions though.

**Comment by davidcok** [comment]
Pick the phrase you want:
- I think the default should brew to say something - if it does not I would get confused as to what had happened
- I’m happy to have a ‘quiet’ option
- Sorry to get the quote wrong, though to me it is still accurate and not confusing

David

File: Source/DafnyDriver/DafnyDriver.cs [hunk 1]
```diff
@@ -384,6 +383,30 @@ public static CommandLineArgumentsResult ProcessCommandLineArguments(string[] ar
       return isVerified;
     }
 
+    private static void WriteTrailer(PipelineStatistics stats) {
+      if (CommandLineOptions.Clo.vcVariety != CommandLineOptions.VCVariety.Doomed && !CommandLineOptions.Clo.Verify && stats.ErrorCount == 0) {
```
Comment (RustanLeino) [inline Source/DafnyDriver/DafnyDriver.cs:387]: Note that there are two flags, the Boogie flag `/noVerify` and the Dafny flag `/dafnyVerify:<n>`. Should this `if` test inspect both?

File: Source/DafnyDriver/DafnyDriver.cs [hunk 2]
```diff
@@ -384,6 +383,30 @@ public static CommandLineArgumentsResult ProcessCommandLineArguments(string[] ar
...
+        if (stats.OutOfMemoryCount != 0) {
+          Console.Write(", {0} out of memory", stats.OutOfMemoryCount);
+        }
+        if (stats.OutOfResourceCount != 0) {
+          Console.Write(", {0} out of resource", stats.OutOfResourceCount);
+        }
+        Console.WriteLine();
+        Console.Out.Flush();
+      } else {
+        // This calls a routine within Boogiemk
```
Comment (RustanLeino) [inline Source/DafnyDriver/DafnyDriver.cs:405]: Boogie

**Comment by RustanLeino** [comment]
I just noticed that before this PR, the behavior is to print "Dafny program verifier finished with 0 verified, 0 errors" with Boogie's `/noVerify` flag. But with Dafny's `/dafnyVerify:0` flag, no message at all gets printed. The latter flag causes the tool to stop after resolution.

File: Source/DafnyDriver/DafnyDriver.cs (hunk 2 above)
Comment (davidcok) [reply Source/DafnyDriver/DafnyDriver.cs]: Fixed typo

File: Source/DafnyDriver/DafnyDriver.cs (hunk 1 above)
Comment (davidcok) [reply Source/DafnyDriver/DafnyDriver.cs:387]: No. If /dafnyVerify:0 is set, control never gets to the point of this informational message.

**Comment by davidcok** [comment]
The behavior with /dafnyVerify:0 is unchanged, namely no output message at all. And perhaps that option can be used for the cases in which Robin would like no output produced.

**Comment by RustanLeino** [comment]
While we're discussing these things, here's something else that many users find confusing about this message: what does "X verified" mean? Technically, the answer is the X is the number of Boogie procedure implementations that (a) had at least 1 proof obligation in the first place, (b) was not excluded by `/proc` or `/noVerify`, and (c) had no errors. This is not particularly meaningful to a Dafny user, not to mention the fact that how Dafny splits proof obligations into Boogie procedures is not known by Dafny users.

Thus, instead of "X verified", it would be nicer to show the count of Dafny declarations for which the verifier was even consulted. (Additional detail: I don't know how this should interact with Boogie's `/proc` flag. I also don't know how this should interact with the fact that Dafny drops any Boogie `implementation` declaration for which there are no proof obligations--case (a) above.)

If we ignore the `/proc`, `/noVerify`, and `/dafnyVerify` options of today and do more redesign, what would a user want/expect to see?

**Comment by RustanLeino** [comment]
> The behavior with /dafnyVerify:0 is unchanged, namely no output message at all. And perhaps that option can be used for the cases in which Robin would like no output produced.

Robin cannot use `/dafnyVerify:0`, because that option never calls the compiler.

**Comment by robin-aws** [comment]
> Robin cannot use `/dafnyVerify:0`, because that option never calls the compiler.

Shoot, got my hopes up there for a minute. That option is even more confusing than what I originally whined about in #442 :)

As for "X verified", I've always found it basically meaningless, and I don't see a way to make it useful even if it was in terms of Dafny constructs. I'm equally uninterested in how many constraints the type checker successfully resolved, for example. :) I imagine Dafny power users or contributors will want some of that detail, but they'll want a lot more detail about WHAT the exact proof obligations were anyway.

> Pick the phrase you want:
> - I think the default should brew to say something - if it does not I would get confused as to what had happened
> - I’m happy to have a ‘quiet’ option
> - Sorry to get the quote wrong, though to me it is still accurate and not confusing

I'm just wondering why the verification stage of the pipeline should do that, whereas you don't get a "Dafny did not compile to a target language" message or a "Dafny also did not compile or run the target language code" message if you tweak the /compile flag.

**Comment by davidcok** [comment]
> 
> I'm just wondering why the verification stage of the pipeline should do that, whereas you don't get a "Dafny did not compile to a target language" message or a "Dafny also did not compile or run the target language code" message if you tweak the /compile flag.
> 
> 
But those generally have artifacts of their own — compilation produces files in the target language; running the program produces output for the execution.

**Comment by robin-aws** [comment]
Returning to this PR, I can live with the noisy output on `/noVerify` for now. I can imagine that being able to suppress that output so that you only get the program output might be useful in the future, IF it is desirable to let the `dafny` CLI run Dafny source directly for that use case. That's a use case we can safely decide to explicitly support in the future, though, and even then silently skipping Dafny verification is probably an anti-pattern anyway!

I'm just going to suggest a tweak to the message and then I think this is fine to go in.

File: Source/DafnyDriver/DafnyDriver.cs [hunk 3]
```diff
@@ -384,6 +383,30 @@ public static CommandLineArgumentsResult ProcessCommandLineArguments(string[] ar
       return isVerified;
     }
 
+    private static void WriteTrailer(PipelineStatistics stats) {
+      if (CommandLineOptions.Clo.vcVariety != CommandLineOptions.VCVariety.Doomed && !CommandLineOptions.Clo.Verify && stats.ErrorCount == 0) {
+        Console.WriteLine();
+        Console.Write("{0} finished without attempting verification", CommandLineOptions.Clo.DescriptiveToolName);
```
Comment (robin-aws) [inline Source/DafnyDriver/DafnyDriver.cs:389]:
```suggestion
        Console.Write("{0} did not attempt verification!", CommandLineOptions.Clo.DescriptiveToolName);
```

**Review by robin-aws — CHANGES_REQUESTED** [review]
Just asking for a wording tweak, otherwise happy for this to go in!

**Review by RustanLeino — APPROVED** [review]

**Review by robin-aws — APPROVED** [review]

**Pull request merged**